### PR TITLE
MGMT-14714: Hypershift agent platform autoscaling

### PR DESCRIPTION
**Pull request by jgyselov**
https://issues.redhat.com/browse/MGMT-14714

Console PR: https://github.com/stolostron/console/pull/3221

### CODE REVIEW

File: libs/ui-lib/lib/cim/components/Hypershift/forms/NodePoolAgentsForm.tsx [hunk 1]
```diff
@@ -29,25 +66,47 @@ const NodePoolAgentsForm = ({
...
         />
       </GridItem>
+      {useAutoscalingValue ? (
+        <>
+          <GridItem>
+            <NumberInputField
+              name={`${autoscalingName}.minReplicas`}
+              label={t('ai:Minimum number of hosts')}
+              isRequired
+              minValue={1}
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/forms/NodePoolAgentsForm.tsx:82]: why dont we allow scaling to 0?

File: libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx [hunk 2]
```diff
@@ -27,13 +27,63 @@ type ManageHostsModalProps = {
...
+      { op: 'remove', path: '/spec/autoScaling' },
+      {
+        op: 'add',
+        value: values.count,
+        path: '/spec/replicas',
+      },
+    );
+  }
+
+  if (values.count !== nodePool.spec.replicas) {
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx:63]: if line 40 or 52 is executed, we will end up with unnecessary patches here. The whole `getPatches` func should be reworked to be more efficient

File: libs/ui-lib/lib/cim/components/Hypershift/modals/NodePoolForm.tsx [hunk 3]
```diff
@@ -24,7 +24,12 @@ export type NodePoolFormValues = {
...
     value: string;
   }[];
-  count: number;
+  count?: number;
+  autoscaling?: {
+    minReplicas: number;
+    maxReplicas: number;
+  };
+  useAutoscaling?: boolean;
 };
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/modals/NodePoolForm.tsx:33]: I'd make all these required & would not allow undefined values

File: libs/ui-lib/lib/cim/components/Hypershift/forms/NodePoolAgentsForm.tsx [hunk 4]
```diff
@@ -29,25 +66,47 @@ const NodePoolAgentsForm = ({
...
-              {t('ai:Maximum availability {{maxAgents}}', { maxAgents })}{' '}
-              <PopoverIcon
-                bodyContent={t(
-                  'ai:Maximum availability is based on the number of hosts available in a given namespace. The number changes dynamically depending on the filtering labels added above.',
-                )}
-              />
-            </>
-          }
+        <CheckboxField
+          name={useAutoscalingName || ''}
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/forms/NodePoolAgentsForm.tsx:70]:
```suggestion
          name={useAutoscalingName}
```

**Review by rawagner — CHANGES_REQUESTED** [review]
we need to add Yup validation that ensures `minReplicas` are less or equal to `maxReplicas`

File: libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx (hunk 2 above)
Comment (jgyselov) [reply libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx]: I hope it's better now :slightly_smiling_face:

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/types.ts [hunk 5]
```diff
@@ -5,12 +5,17 @@ import { NodePoolK8sResource } from '../../types';
...
     key: string;
     value: string;
   }[];
   releaseImage: string;
+  count?: number;
+  autoscaling?: {
+    minReplicas: number;
+    maxReplicas: number;
+  };
+  useAutoscaling?: boolean;
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/types.ts:18]:
```suggestion
  count: number;
  autoscaling: {
    minReplicas: number;
    maxReplicas: number;
  };
  useAutoscaling: boolean;
```

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/HostsForm.tsx [hunk 6]
```diff
@@ -35,7 +35,7 @@ const HostsForm: React.FC<HostsFormProps> = ({
       ];
 
   const totalHosts = values.nodePools.reduce((acc, nodePool) => {
-    acc += nodePool.count;
+    acc += (nodePool.count || nodePool.autoscaling?.maxReplicas) as number;
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/HostsForm.tsx:38]:
```suggestion
    acc += nodePool.useAutoscaling ? nodePool.autoscaling.maxReplicas : nodePool.count;
```

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx [hunk 7]
```diff
@@ -63,20 +69,44 @@ const NodePoolForm: React.FC<NodePoolFormProps> = ({ infraEnvs, agents, index, o
   let previousNodePoolsCount = 0;
 
   for (let i = 0; i < index; i++) {
-    previousNodePoolsCount += values.nodePools[i].count;
+    previousNodePoolsCount += (
+      values.nodePools[i].useAutoscaling
+        ? values.nodePools[i].autoscaling?.maxReplicas
+        : values.nodePools[i].count
+    ) as number;
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx:76]:
```suggestion
    previousNodePoolsCount += (
      values.nodePools[i].useAutoscaling
        ? values.nodePools[i].autoscaling.maxReplicas
        : values.nodePools[i].count
    );
```

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx [hunk 8]
```diff
@@ -63,20 +69,44 @@ const NodePoolForm: React.FC<NodePoolFormProps> = ({ infraEnvs, agents, index, o
...
     0,
     Math.min(matchingAgents.length, availableAgents.length - previousNodePoolsCount),
   );
 
-  const currentCount = values.nodePools[index].count;
+  const currentCount = (
+    values.nodePools[index].useAutoscaling
+      ? values.nodePools[index].autoscaling?.maxReplicas
+      : values.nodePools[index].count
+  ) as number;
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx:88]:
```suggestion
  const currentCount = (
    values.nodePools[index].useAutoscaling
      ? values.nodePools[index].autoscaling.maxReplicas
      : values.nodePools[index].count
  );
```

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx [hunk 9]
```diff
@@ -63,20 +69,44 @@ const NodePoolForm: React.FC<NodePoolFormProps> = ({ infraEnvs, agents, index, o
...
+
   React.useEffect(() => {
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx:90]: this `useEffect` has a few issues
 - it does not check `useAutoscaling` to determine which form value to update
 - the whole autoscaling branch is flawed (ie setting min value only if max agents === 0, but what about max being less than min)
 - generally I think this useEffect which is changing values without notifying user was not a good decision

We should fix the first two. The third one, which is a bigger change, can be deferred.

File: libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx [hunk 10]
```diff
@@ -27,13 +27,60 @@ type ManageHostsModalProps = {
...
+const getPatches = (values: NodePoolFormValues, nodePool: NodePoolK8sResource) => {
+  const patches = [];
+
+  if (values.useAutoscaling) {
+    if (!!nodePool.spec.autoScaling) {
+      patches.push({
+        op: 'replace',
+        value: {
+          min: values.autoscaling?.minReplicas,
+          max: values.autoscaling?.maxReplicas,
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx:46]:
```suggestion
          min: values.autoscaling.minReplicas,
          max: values.autoscaling.maxReplicas,
```

File: libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx [hunk 11]
```diff
@@ -27,13 +27,60 @@ type ManageHostsModalProps = {
...
+        path: '/spec/autoScaling',
+      });
+    } else {
+      patches.push(
+        { op: 'remove', path: '/spec/replicas' },
+        {
+          op: 'add',
+          value: {
+            min: values.autoscaling?.minReplicas,
+            max: values.autoscaling?.maxReplicas,
```
Comment (rawagner) [inline libs/ui-lib/lib/cim/components/Hypershift/modals/ManageHostsModal.tsx:57]:
```suggestion
            min: values.autoscaling.minReplicas,
            max: values.autoscaling.maxReplicas,
```

**Review by rawagner — CHANGES_REQUESTED** [review]
Just a few more enhancements please :)

File: libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx (hunk 9 above)
Comment (jgyselov) [reply libs/ui-lib/lib/cim/components/Hypershift/HostedClusterWizard/HostsStep/NodePoolForm.tsx:86]: I fixed the first two. I also added validation on the autoscaling fields that triggers an error when the numbers are <1.

image

**Review by rawagner — DISMISSED** [review]

**Review by ammont82 — DISMISSED** [review]

**Review by rawagner — APPROVED** [review]

**Review by ammont82 — APPROVED** [review]

**Pull request merged**